hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/StateContext.java

```diff
@@ -190,16 +211,29 @@ void setShutdownGracefully() {
   public boolean getShutdownOnError() {
     return shutdownOnError;
   }
+
   /**
    * Adds the report to report queue.
    *
    * @param report report to be added
    */
   public void addReport(GeneratedMessage report) {
     if (report != null) {
-      synchronized (reports) {
-        for (InetSocketAddress endpoint : endpoints) {
-          reports.get(endpoint).add(report);
+      // TODO: Check report.getDescriptorForType() != null as well?
+      final String reportType = report.getDescriptorForType().getFullName();
+      for (InetSocketAddress endpoint : endpoints) {
+        // We only keep the latest container, node and pipeline report
+        if (reportType.equals(CONTAINER_REPORTS_PROTO_NAME)) {
```

Comment:
Do we need to synchronize the get/set on these containerReport/nodeReport/pipelineReport ?

Reply:
No container/node/pipeline should be fine. We need `synchroized` on `reports` because we want Map `get` and `add` not to get messed in between.

Reply:
one thread write containerReport and another thread read containerReport, it's not safe, we can use `AtomicReference<GeneratedMessage> containerReport`

Reply:
done